design/thin-edge-core.md

```diff
@@ -0,0 +1,50 @@
+# Thin-edge Core Vision
+
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools, actuators, cloud end-points;
+* __constraint environments__ with low-resources devices, large-scale deployments, high exposure to security risks,
+  prohibitive manual operations, long-term supports;
+* __safety related requirements__ with devices operating machines in hazardous contexts.
+  
+The forces being these requirements are pushing in opposite directions.
+
+* On one side, one needs a dynamic and open setting with a large set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, one needs a minimal system specifically built for a task with cherry-picked components.
+  On the extreme side this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel) and sealed, fixed-purpose software images.
```

Comment:
General comment here IMHO, that's why the unikernels are not that widely used. The biggest problem is to port the existing applications into these specific kernels. But I do remember that the work is going to ease this by porting the libc kind of libraries into these unikernels, so that the existing applications can run easily on these kernels. One such example is [here ](https://github.com/cloudius-systems/osv)

Reply:
Unikernels are pushing the approach to the extreme and are therefore complex to manage, sure. However, I'm not advocating to go so extreme, but to provide the tools for our customers to be able to build smaller and specific executables. By contrast, with the current implementation, the device owner is given some flexibility on runtime (say to connect Azure and not Cumulocity), but the executables embark *all* the mappers even if you plan to never use them. The impacts are not only on the code size, but also leading to useless updates, side-effect bugs, larger security surface. A customer, who plans to never use a mapper, should not be impacted by this mapper.